src/components/Photo.js

```diff
@@ -0,0 +1,24 @@
+import axios from 'axios';
+import React, { useEffect, useState } from 'react';
+
+const serverUrl = process.env.SERVER_URL;
+
+const Photo = (props) => {
+  const [ src, setSrc ] = useState(null);
+
+  // Similar to componentDidMount and componentDidUpdate:
+  useEffect(() => {
+    axios(`http://${serverUrl}/assets/${props.src}`)
+      .then((result) => {
+        setSrc(result.data);
+      });
+  });
+
+  return (
+    <div>
+      <img src={src} />
```

Comment:
Mind helping me understand this flow? I think my understanding is that this is the construction of a `Photo` object, useEffect and useState are some functions from react, and somehow from useState we get the src const(?) and a setSrc function, which we call to assign to the src var/obj(?) This feels very indirect but I understand that this might just be how js works in doing async stuff to a remote server and then updating UI after.

Reply:
You have it pretty much exactly right: 

### `useState`
Returns two variables `src` which is the actual variable, and `setSrc` which is a function that updates the `src` variable. 

This basically tells react: "I want to keep state on this variable, and i'm going to use that function to update it".

### `useEffect`
This is a bit trickier. Basically runs as soon as the component loads, so at this point in time, we don't have the image url because it needs to be signed, and to sign it we need to be auth'd. So what we do is basically say: "as soon as the component loads, fire this request to the server to get the signed url. Once we get a response from the server, then please update the source using `setSrc(response.data)`. 

Part of the `setSrc` under the hood functionality is that it sees it's being called and then immediately re-renders the component, except now the `src` variable is actually loaded with the right path. 

How this all comes together is that we can basically lazily wait for the server to return the url and when we get it, then render the right thing on the screen. With this flow, we could do something like have a temporary thumbnail in the tag before we render the real thing with some super simple logic: 

```
<img src={src || 'loading.gif'} />
```

Reply:
This all works on promises, which is a tricky but pretty powerful concept. It allows the component to render, not holding the page hostage to the server, and update only the small specific elements in the page as the data comes from the server. 

In this cause, if we had a gallery with 100 images, we don't have to wait for 100 calls to the server... instead, one by one, as the responses come, each image will pop up into view.